cmd/scheduleapi/handler/get_arrival_estimates.go

```diff
@@ -0,0 +1,186 @@
+package handler
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/request"
+	"github.com/aws/aws-sdk-go/service/dynamodb"
+	"github.com/aws/aws-sdk-go/service/dynamodb/dynamodbattribute"
+	"github.com/aws/aws-sdk-go/service/dynamodb/expression"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/pkg/errors"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/smartatransit/fivepoints/api/v1/schedule"
+	"github.com/smartatransit/fivepoints/pkg/martaapi"
+)
+
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 . DynamoQuerier
+type DynamoQuerier interface {
+	QueryPagesWithContext(ctx aws.Context, input *dynamodb.QueryInput, fn func(*dynamodb.QueryOutput, bool) bool, opts ...request.Option) error
+	QueryWithContext(ctx aws.Context, input *dynamodb.QueryInput, opts ...request.Option) (*dynamodb.QueryOutput, error)
+}
+
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 . GetArrivalEstimatesEndpoint
+type GetArrivalEstimatesEndpoint func(context.Context, *schedule.GetArrivalEstimatesRequest) (*schedule.GetArrivalEstimatesResponse, error)
+
+func ValidateRequest(ctx context.Context, in *schedule.GetArrivalEstimatesRequest) error {
+	var errStrings []string
+	if in == nil {
+		errStrings = append(errStrings, "request body is nil")
+	}
+	if strings.TrimSpace(in.GetDestination()) == "" {
+		errStrings = append(errStrings, "destination is nil")
+	}
+	if strings.TrimSpace(in.GetStation()) == "" {
+		errStrings = append(errStrings, "station is nil")
+	}
+	if in.GetStartDate() == nil {
+		errStrings = append(errStrings, "start date is nil")
+	}
+	if in.GetEndDate() == nil {
+		errStrings = append(errStrings, "end date is nil")
+	}
+	if in.GetStartDate().GetSeconds() > in.GetEndDate().GetSeconds() {
+		errStrings = append(errStrings, "start date must be before end date")
+	}
+	s, err := ptypes.Timestamp(in.GetStartDate())
+	if err != nil {
+		errStrings = append(errStrings, "start must be RFC3339 encoded")
```

Comment:
I think that the `timestamp.Timstamp` that you're converting into a `time.Time` here is actually a pair of `int64`s rather than a string representation. This should only fail if they're outside of some arbitrary range that protobuf has decided is the legal range for their timestamps https://github.com/golang/protobuf/blob/master/ptypes/timestamp.go#L90

Reply:
This is a good point!  
Looking into this, I think if given a valid `RFC3339`, start and end will always be in a valid range:
https://github.com/golang/protobuf/blob/master/ptypes/timestamp/timestamp.pb.go#L29-L33

so I think this still might be a valid claim, since if it was outside of that range then protobuf would have encoded it differently?

Reply:
well all I mean is that this endpoint doesn't accept a string at all, so it's not clear what the error message is referring to. The `grpc-gateway` will presumably receive a string and attempt to convert it to a pbuf type, but if the timestamp were malformatted,  _that's_ where the error would originate. The grpc endpoint, rather, accepts a pair of `int64`s, and my feeling is that it should complain about `int64`s and not strings if something is wrong, but this is a mega nitpick